### PR TITLE
Move embedding to ATen

### PR DESCRIPTION
**Pull request by colesbury**
Breaking change: new optional argument padding_idx in F.embedding to
match nn.Embedding.

Note that there are a few bugs in Embedding that are inherited from the
previous code:

 - CUDA renorm has race conditions if index contains duplicate entries
 - sparse gradient doesn't work with scale_grad_by_freq

### CODE REVIEW

File: aten/src/ATen/native/Embedding.cpp [hunk 1]
```diff
@@ -0,0 +1,181 @@
...
+  auto indices_arg = TensorArg(indices, "indices", 2);
+  checkContiguous("embedding_renorm_", self_arg);
+  checkDim("embedding_renorm_", self_arg, 2);
+  checkContiguous("embedding_renorm_", indices_arg);
+  checkScalarType("embedding_renorm_", indices_arg, kLong);
+
+  auto num_indices = indices.numel();
+  auto indices_data = std::unique_ptr<int64_t[]>(new int64_t[num_indices]);
+  std::memcpy(indices_data.get(), indices.data_ptr(), num_indices * sizeof(int64_t));
+  std::qsort(indices_data.get(), num_indices, sizeof(int64_t),
```
Comment (apaszke) [inline aten/src/ATen/native/Embedding.cpp]: Isn't `std::sort` more efficient than `std::qsort` (don't think the compiler can inline the check as a lambda, but `std::sort` is a template, so it's inlined for sure). Just use `std::vector` for indices, and take its `.data()`

File: aten/src/ATen/Dispatch.h [hunk 2]
```diff
@@ -42,5 +42,22 @@ auto dispatch_floating_types(const Type& the_type, const char *name, Args&&... a
     }
 }
 
+#define DISPATCH_ALL_FLOATING_TYPES(TYPE, NAME, F...)                            \
+  [&]() {                                                                         \
```
Comment (gchanan) [inline aten/src/ATen/Dispatch.h]: I'm not the biggest fan of this style if we don't have C++14, because it has some downsides that are eliminated with the existing template style at the cost of ~2-3 lines of code:

1) I don't think you can capture things as r-value references this way in C++11 (you can in C++14), meaning (among other things) you can't capture unique_ptrs.
2) "scalar" is magical (I need to know the definition, not just that it dispatches to my templatized function) and might shadow your own definition in non-obvious ways.  If we had C++14  we could use a templatized lambda and avoid the problem.

File: aten/src/ATen/native/cuda/AccumulateType.h [hunk 3]
```diff
@@ -0,0 +1,22 @@
+#pragma once
```
Comment (gchanan) [inline aten/src/ATen/native/cuda/AccumulateType.h]: In https://github.com/pytorch/pytorch/pull/4304, I started putting utility-style cuda code in /cuda rather than /native/cuda.  I think this makes sense because technically everything that isn't a TH/THC wrapper is "native", so we need to have a more precise definition of native to limit what goes in /native; I'd argue functional extensions are a natural definition.  Since this seems generally useful and isn't just implementing a functional extension, I'd put it in /cuda.

File: torch/nn/_functions/thnn/sparse.py [hunk 4]
```diff
@@ -6,120 +6,6 @@
 from . import _all_functions
 
 
-class Embedding(Function):
-
-    @staticmethod
-    def symbolic(g, indices, weight, padding_idx, max_norm, norm_type, scale_grad_by_freq,
```
Comment (gchanan) [inline torch/nn/_functions/thnn/sparse.py]: do you need to move this?

File: aten/src/ATen/native/Embedding.cpp [hunk 5]
```diff
@@ -0,0 +1,181 @@
...
+#endif
+
+
+namespace at { namespace native {
+
+Tensor embedding(const Tensor & weight, const Tensor & indices,
+                 int64_t padding_idx, bool scale_grad_by_freq, bool sparse) {
+  auto indices_arg = TensorArg(indices, "indices", 1);
+  checkScalarType("embedding", indices_arg, kLong);
+  checkContiguous("embedding", indices_arg);
```
Comment (gchanan) [inline aten/src/ATen/native/Embedding.cpp]: previously, we would set these to contiguous if they weren't; should we do that in the functional call for backwards compat?

File: aten/src/ATen/native/cuda/AccumulateType.h [hunk 6]
```diff
@@ -0,0 +1,22 @@
...
+#include <cuda.h>
+#include <cuda_fp16.h>
+
+namespace at { namespace native { namespace cuda {
+
+template <typename T>
+struct AccumulateType { using type = T; };
+
+template <>
+struct AccumulateType<half> { using type = float; };
```
Comment (gchanan) [inline aten/src/ATen/native/cuda/AccumulateType.h]: why are the int types not specified as they are in THC?  (I know you don't use them, but if someone else starts using them they'll screw things up).

File: aten/src/ATen/Dispatch.h (hunk 2 above)
Comment (colesbury) [reply aten/src/ATen/Dispatch.h]: 1. Not sure what you're referring to. You can capture `unique_ptr` by reference.
2. It's not much more magical than assuming your templated struct takes a the scalar type as the first parameter. It just standardizes the name.

C++14 won't help. We don't really want lambdas with `auto` here because the lambdas don't typically take anything of type `scalar`. They just take `Tensor`s.

It's substantially less boiler plate because you can capture the relevant tensors by reference. With a struct functor you have to pass all argument through your lambda, as well as possibly the grid, dim, and stream.

File: aten/src/ATen/native/cuda/AccumulateType.h (hunk 3 above)
Comment (colesbury) [reply aten/src/ATen/native/cuda/AccumulateType.h]: Yeah, I like that better. I'll rebase and move it there after you merge your PR.

File: aten/src/ATen/native/cuda/AccumulateType.h (hunk 6 above)
Comment (colesbury) [reply aten/src/ATen/native/cuda/AccumulateType.h]: Good point. I'll add them.

Comment (gchanan) [reply aten/src/ATen/native/cuda/AccumulateType.h]: It's also nice that we can see the definitions all in one place instead of having to look through 10 files...

File: aten/src/ATen/Dispatch.h (hunk 2 above)
Comment (gchanan) [reply aten/src/ATen/Dispatch.h]: grr...it looks like it's a C++20 feature...

File: torch/nn/_functions/thnn/sparse.py (hunk 4 above)
Comment (colesbury) [reply torch/nn/_functions/thnn/sparse.py]: Good catch. Done.

File: aten/src/ATen/native/Embedding.cpp (hunk 5 above)
Comment (colesbury) [reply aten/src/ATen/native/Embedding.cpp]: I'll add it to the Python code `def embedding`

File: aten/src/ATen/native/cuda/Embedding.cu [hunk 7]
```diff
@@ -0,0 +1,348 @@
...
+  const int SZ = 4;
+
+  if (idx < numel
+      && (idx == 0 || input[idx] != input[idx - 1])
+      && input[idx] != padding_idx) {
+    do {
+      const int start_feature = threadIdx.x + blockIdx.y * blockDim.x * SZ;
+      const int weight_row = ((int) input[idx]) * stride;
+      const int grad_row = ((int) indices[idx]) * stride;
+      const accscalar_t scale = count ? (accscalar_t)1.0 / count[idx] : 1.0;
```
Comment (gchanan) [inline aten/src/ATen/native/cuda/Embedding.cu]: why do you need the conversion on only one of the cases? (there's no warning on the false case?)

**Review by gchanan — APPROVED** [review]
Awesome to get this into C++.  Ideally we wouldn't THC stuff directly in here (we'd have wrappers), so folks only need to learn one API, but this is big enough already.

File: aten/src/ATen/Dispatch.h [hunk 8]
```diff
@@ -42,5 +42,22 @@ auto dispatch_floating_types(const Type& the_type, const char *name, Args&&... a
     }
 }
 
+#define DISPATCH_ALL_FLOATING_TYPES(TYPE, NAME, F...)                             \
+  [&]() {                                                                         \
+    const Type& the_type = TYPE;                                                  \
```
[the remark this answers is not in the document]
Comment (colesbury) [reply aten/src/ATen/Dispatch.h]: This ensures that if you do something like `DISPATCH_ALL_FLOATING_TYPES(self, ...)` instead of `DISPATCH_ALL_FLOATING_TYPES(self.type(), ...)` you'll get a meaningful compile error early on.

I'm not too worried about another variable in-scope. It can be shadowed in `F` without generating compile warnings.

[the remark this answers is not in the document]
Comment (gchanan) [reply aten/src/ATen/Dispatch.h]: up to you, I don't like people messing with my scope :).  You can get the same behavior by just enclosing the assignment in brackets and using TYPE from then on.

File: aten/src/ATen/Dispatch.h [hunk 9]
```diff
@@ -44,5 +44,22 @@ R dispatch_floating_types(const Type& the_type, const char *name, Args&&... args
   }
 }
 
+#define DISPATCH_ALL_FLOATING_TYPES(TYPE, NAME, F...)                             \
```
Comment (yf225) [inline aten/src/ATen/Dispatch.h]: Windows CI wasn't working properly earlier, but it seems that MSVC complains about `error C2010: '.': unexpected in macro formal parameter list`. Any suggestions on how to fix it?

Comment (colesbury) [reply aten/src/ATen/Dispatch.h]: I think we can replace `F...` with `...` and `F()` with `__VA_ARGS__()`

File: aten/src/ATen/native/cuda/Embedding.cu [hunk 10]
```diff
@@ -0,0 +1,348 @@
...
+    } else {
+      v += std::pow(x, norm_type);
+    }
+  }
+
+  using Op = ReduceAdd<accscalar_t, accscalar_t>;
+  v = reduceBlock<accscalar_t>(sdata, blockDim.x, v, Op(), 0);
+
+  if (tid == 0) {
+    sdata[0] = std::pow(v, 1.0 / scalar_cast<accscalar_t>(norm_type));
```
Comment (yf225) [inline aten/src/ATen/native/cuda/Embedding.cu]: MSVC seems to complain about the following:
```
error: calling a __host__ function("pow<float, double, void> ") from a __global__ function("at::native::_NV_ANON_NAMESPACE::renorm_kernel<float, float> ") is not allowed
error: identifier "pow<float, double, void> " is undefined in device code
error: calling a __host__ function("pow<float, double, void> ") from a __global__ function("at::native::_NV_ANON_NAMESPACE::renorm_kernel< ::__half, float> ") is not allowed
error: identifier "pow<float, double, void> " is undefined in device code
```

I tried to add `#include "math.h"` at the top of file to include the CUDA math library, but it seems that calling `pow(...)` inside the `at::native` namespace will be calling the ATen tensor `pow(...)` instead.
